client/client.go

```diff
@@ -396,39 +253,132 @@ type client struct {
-func NewClientWithContext(ctx context.Context, pdAddrs []string, security SecurityOption, opts ...ClientOption) (Client, error) {
-	log.Info("[pd] create pd client with endpoints", zap.Strings("pd-address", pdAddrs))
+func NewClientWithContext(ctx context.Context, svrAddrs []string, security SecurityOption, opts ...ClientOption) (Client, error) {
+	log.Info("[pd] create pd client with endpoints", zap.Strings("pd-address", svrAddrs))
+	c, clientCtx, clientCancel, tlsCfg := createClient(ctx, &security)
+	c.tsoStreamBuilderFactory = &pdTSOStreamBuilderFactory{}
+	c.bc = newPDBaseClient(clientCtx, clientCancel, &c.wg, addrsToUrls(svrAddrs), tlsCfg, c.option)
+	if err := c.setup(true, true, opts...); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// NewTSOClientWithContext creates a TSO client with context.
+func NewTSOClientWithContext(ctx context.Context, svrAddrs []string, security SecurityOption, opts ...ClientOption) (Client, error) {
+	log.Info("[pd(tso)] create tso client with endpoints", zap.Strings("tso-address", svrAddrs))
+	c, clientCtx, clientCancel, tlsCfg := createClient(ctx, &security)
+	c.tsoStreamBuilderFactory = &tsoTSOStreamBuilderFactory{}
+	c.bc = newTSOBaseClient(clientCtx, clientCancel, &c.wg, addrsToUrls(svrAddrs), tlsCfg, c.option)
+	if err := c.setup(true, false, opts...); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func createClient(ctx context.Context, security *SecurityOption) (*client, context.Context, context.CancelFunc, *tlsutil.TLSConfig) {
+	tlsCfg := &tlsutil.TLSConfig{
+		CAPath:   security.CAPath,
+		CertPath: security.CertPath,
+		KeyPath:  security.KeyPath,
+
+		SSLCABytes:   security.SSLCABytes,
+		SSLCertBytes: security.SSLCertBytes,
+		SSLKEYBytes:  security.SSLKEYBytes,
+	}
+
+	clientCtx, clientCancel := context.WithCancel(ctx)
 	c := &client{
-		baseClient:        newBaseClient(ctx, addrsToUrls(pdAddrs), security),
-		checkTSDeadlineCh: make(chan struct{}),
+		checkTSDeadlineCh:         make(chan struct{}),
+		checkTSODispatcherCh:      make(chan struct{}, 1),
+		updateTSOConnectionCtxsCh: make(chan struct{}, 1),
+		updateTokenConnectionCh:   make(chan struct{}, 1),
+		ctx:                       clientCtx,
+		cancel:                    clientCancel,
+		option:                    newOption(),
 	}
+
+	return c, clientCtx, clientCancel, tlsCfg
+}
+
+func (c *client) setup(enableTSO, enableAdmissionCtl bool, opts ...ClientOption) error {
 	// Inject the client options.
 	for _, opt := range opts {
 		opt(c)
 	}
 	// Init the client base.
-	if err := c.init(); err != nil {
-		return nil, err
+	if err := c.bc.Init(); err != nil {
+		return err
+	}
+
+	// Register callbacks and start the daemons.
+	if enableTSO {
+		c.bc.AddTSOAllocatorServingAddrSwitchedCallback(c.scheduleCheckTSODispatcher)
+		c.bc.AddServiceAddrsSwitchedCallback(c.scheduleUpdateTSOConnectionCtxs)
+		c.updateTSODispatcher()
+		c.wg.Add(2)
+		go c.tsLoop()
+		go c.tsCancelLoop()
+	}
+	if enableAdmissionCtl {
```

Comment:
what does it mean?

Reply:
tso mcs mode doesn't need to enable admisstion control module, so this flag is needed.

Reply:
Do you mean we can have a better name, such as enableAdmissionControl?

Reply:
I remove the flags for now. The original purpose of adding these flags is to disable some clients when the server side doesn't provide the correspoinding services. For example, disable tso client when the server side is resource management microservice, or disable resource management client when the server side the tso microservice.